otherlibs/fiber/pool.ml

```diff
@@ -20,26 +30,59 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { mvar = Mvar.create (); status = Open }
+let create () =
+  { tasks = Queue.create (); runner = Awaiting_run; status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
-  | Open -> Mvar.write t.mvar (Task f) k
-
-let stream t =
-  Stream.In.create (fun () ->
-      let+ next = Mvar.read t.mvar in
-      match next with
-      | Done -> None
-      | Task task -> Some task)
+  | Open -> (
+    Queue.push t.tasks f;
+    match t.runner with
+    | Running | Awaiting_run -> k ()
+    | Awaiting_resume r ->
+      t.runner <- Running;
+      resume r () k)
 
 let stop t k =
   match t.status with
   | Closed -> k ()
-  | Open ->
+  | Open -> (
     t.status <- Closed;
-    Mvar.write t.mvar Done k
+    match t.runner with
+    | Running | Awaiting_run -> k ()
+    | Awaiting_resume r ->
+      t.runner <- Running;
+      resume r () k)
 
-let run t = stream t |> Stream.In.parallel_iter ~f:(fun task -> task ())
+let run t k =
+  match t.runner with
+  | Awaiting_resume _ | Running ->
+    Code_error.raise "Fiber.Pool.run: concurent calls to run aren't allowed" []
+  | Awaiting_run ->
+    t.runner <- Running;
+    (* The number of currently running fibers in the pool. Only when this
+        number reaches zero we may call the final continuation [k]. *)
+    let n = ref 1 in
+    let done_fiber () =
+      decr n;
+      if !n = 0 then k () else end_of_fiber
+    in
+    let rec read t =
+      match Queue.pop t.tasks with
+      | None -> on_finish t
+      | Some v ->
+        incr n;
+        fork (fun () -> v () done_fiber) read_delayed
+    and read_delayed () = read t
+    and suspend_k k =
+      (* we are suspending because we have no tasks *)
+      assert (Queue.is_empty t.tasks);
+      t.runner <- Awaiting_resume k
+    and on_finish t =
```

Comment:
This isn't really "finishing" in any sense, it's more like "start_waiting" instead, right?

Reply:
I called it `suspend_or_finish`. It only waits if it's not closed. If it's closed, it will indeed "finish"